components/style/restyle_hints.rs

```diff
@@ -1180,97 +1188,108 @@ impl DependencySet {
-            let mut now_context =
-                MatchingContext::new_for_visited(MatchingMode::Normal, bloom_filter,
-                                                 VisitedHandlingMode::AllLinksUnvisited,
-                                                 shared_context.quirks_mode);
-            let matches_now =
-                matches_selector(&dep.selector,
-                                 dep.selector_offset,
-                                 &dep.hashes,
-                                 el,
-                                 &mut now_context,
-                                 &mut |_, _| {});
-
-            // Check for mismatches in both the match result and also the status
-            // of whether a relevant link was found.
-            if matched_then != matches_now ||
-               then_context.relevant_link_found != now_context.relevant_link_found {
-                hint.insert_from(&dep.hint);
-                return !hint.is_maximum()
-            }
+                if hint.contains(&dep.hint) {
+                    trace!(" > hint was already there");
+                    return true;
+                }
 
-            // If there is a relevant link, then we also matched in visited
-            // mode.  Match again in this mode to ensure this also matches.
-            // Note that we never actually match directly against the element's
-            // true visited state at all, since that would expose us to timing
-            // attacks.  The matching process only considers the relevant link
-            // state and visited handling mode when deciding if visited
-            // matches.  Instead, we are rematching here in case there is some
-            // :visited selector whose matching result changed for some _other_
-            // element state or attribute.
-            if now_context.relevant_link_found &&
-               dep.sensitivities.states.intersects(IN_VISITED_OR_UNVISITED_STATE) {
-                then_context.visited_handling = VisitedHandlingMode::RelevantLinkVisited;
+                // NOTE(emilio): We can't use the bloom filter for snapshots, given
+                // that arbitrary elements in the parent chain may have mutated
+                // their id's/classes, which means that they won't be in the
+                // filter, and as such we may fast-reject selectors incorrectly.
+                //
+                // We may be able to improve this if we record as we go down the
+                // tree whether any parent had a snapshot, and whether those
+                // snapshots were taken due to an element class/id change, but it's
+                // not clear we _need_ it right now.
+                let mut then_context =
+                    MatchingContext::new_for_visited(MatchingMode::Normal, None,
+                                                     VisitedHandlingMode::AllLinksUnvisited,
+                                                     shared_context.quirks_mode);
```

Comment:
I'm assuming all this is pretty much a re-indent, plus passing the right case sensitivity, right? If it isn't please comment on which other changes are there, because it may matter (I'm rewriting this in https://bugzilla.mozilla.org/show_bug.cgi?id=1368240)